morphia/src/main/java/org/mongodb/morphia/utils/ReflectionUtils.java

```diff
@@ -558,9 +566,13 @@ public static Entity getClassEntityAnnotation(final Class c) {
                 jarEntry = jarFile.getNextJarEntry();
                 if (jarEntry != null) {
                     String className = jarEntry.getName();
-                    if (className.endsWith(".class") && getPackageName(className).equals(packageName)) {
-                        className = stripFilenameExtension(className);
-                        classes.add(Class.forName(className.replace('/', '.'), true, loader));
+                    if (className.endsWith(".class")) {
+                        String classPackageName = getPackageName(className);
+                        if ((ignoreSubPackages && classPackageName.equals(packageName)) || (!ignoreSubPackages && isSubPackage(
+                                classPackageName, packageName))) {
```

Comment:
the first clause need only check equals().  checking ignoreSubPackages twice is a bit redundant.